Cabal/Distribution/Parsec/ParseResult.hs

```diff
@@ -1,6 +1,14 @@
 {-# LANGUAGE BangPatterns #-}
 {-# LANGUAGE CPP          #-}
 {-# LANGUAGE RankNTypes   #-}
+
+-- For bootstrapping with older GHCs (#4785)
+#ifdef MIN_VERSION_base
+#if MIN_VERSION_base(4,10,0)
+#define HAVE_base_4_10
+#endif
+#endif
```

Comment:
I'd feel better if we had at least a `#warning falling back to heuristic mode due to cabal-install < 1.24 && GHC < 8` in an `#else` branch so it's clear that we're in uncertain territory.

Reply:
See e.g. https://github.com/ekmett/lens/blob/master/Setup.lhs#L25-L32 for a common boilerplate warning

Reply:
I'm not opposed to that in principle at all but I will not that there are already quite a few places in the code where we do things like this. Should I patch those as well? Maybe it would be better to collect all these `#ifdefs` into a single header file and include that where these are needed?

Reply:
I know I've seen `#warning` boilerplate in some Setup.hs but I can't find it rightaway, maybe @phadej knows an example... :-)

Reply:
Right, there should be only one warning, in `Setup.hs`.

Reply:
I was thinking of defining the `HAVE_` macros in a central header file and generate the warning there.

Reply:
@DanielG For some reason I'm a bit worried about introducing an `#include` file as it would require search paths to set up, potentially breaking `ghc --make Setup.hs`

Reply:
Fair enough. Personally I think all CPP is code smell but what're you gonna do ;)

I've been thinking we need a proper Haskell preprocessor for a while now.. `{-# LANGUAGE HPP #-}` anyone?

Reply:
> I've been thinking we need a proper Haskell preprocessor for a while now.. `{-# LANGUAGE HPP #-}` anyone?

Yes! A first step would be https://ghc.haskell.org/trac/ghc/wiki/Proposal/NativeCpp ... if you want to work on that, lemme know... :-)